### PR TITLE
[SCM-885] second implementation with new API

### PR DESCRIPTION
**Pull request by McFoggy**
this PR is a fix for SCM-885 which keeps the git exe implementation backard compatible but thus introduces a new command parameter: START_FROM_ROOT

the PR also adds several TCK tests and provides corrections of existing faulty behaviors of the jgit implementation

### CODE REVIEW

**Comment by michael-o** [comment]
Please revise this PR for the conflict.

File: maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/AbstractChangeLogCommand.java [hunk 1]
```diff
@@ -50,11 +50,20 @@ protected abstract ChangeLogScmResult executeChangeLogCommand( ScmProviderReposi
     protected ChangeLogScmResult executeChangeLogCommand( ScmProviderRepository repository, ScmFileSet fileSet,
                                                           ScmVersion startVersion, ScmVersion endVersion,
                                                           String datePattern )
-        throws ScmException
+            throws ScmException
```
Comment (michael-o) [inline maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/AbstractChangeLogCommand.java]: Drop this change.

File: maven-scm-api/src/main/java/org/apache/maven/scm/CommandParameter.java [hunk 2]
```diff
@@ -50,6 +50,8 @@
 
     public static final CommandParameter BRANCH = new CommandParameter( "branch" );
 
+    public static final CommandParameter START_FROM_ROOT = new CommandParameter( "startFromRoot" );
```
Comment (michael-o) [inline maven-scm-api/src/main/java/org/apache/maven/scm/CommandParameter.java]: Why did you call it `root`. Looking at [gitvevisions](https://git-scm.com/docs/gitrevisions#_revision_range_summary) I fail to find a corresponding term. Does `startFromEnd` make more sense?

File: maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/AbstractChangeLogCommand.java [hunk 3]
```diff
@@ -50,11 +50,20 @@ protected abstract ChangeLogScmResult executeChangeLogCommand( ScmProviderReposi
...
         throw new ScmException( "Unsupported method for this provider." );
     }
 
+    protected ChangeLogScmResult executeChangeLogCommand( ScmProviderRepository repository, ScmFileSet fileSet,
+                                                          ScmVersion endVersion,
+                                                          String datePattern, boolean startFromRoot )
+            throws ScmException
+    {
+        // by default let's just delegate to executeChangeLogCommand with both start and end versions
+        return executeChangeLogCommand( repository, fileSet, null, endVersion, datePattern );
```
Comment (michael-o) [inline maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/AbstractChangeLogCommand.java]: What is the purpose of this entire method if `startFromRoot` is not passed?

File: maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/ChangeLogScmRequest.java [hunk 4]
```diff
@@ -130,6 +130,18 @@ public void setEndRevision( ScmVersion endRevision )
         parameters.setScmVersion( CommandParameter.END_SCM_VERSION, endRevision );
     }
 
+    public boolean getStartFromRoot()
```
Comment (michael-o) [inline maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/ChangeLogScmRequest.java]: Boolean return value should be `is...`.

File: maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/ChangeLogScmRequest.java [hunk 5]
```diff
@@ -130,6 +130,18 @@ public void setEndRevision( ScmVersion endRevision )
...
+        throws ScmException
```
Comment (michael-o) [inline maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/ChangeLogScmRequest.java]: 4 char indent

File: maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/ChangeLogScmRequest.java [hunk 6]
```diff
@@ -130,6 +130,18 @@ public void setEndRevision( ScmVersion endRevision )
...
+    {
+        return parameters.getBoolean( CommandParameter.START_FROM_ROOT, false );
+    }
+
+    public void setStartFromRoot()
+            throws ScmException
```
Comment (michael-o) [inline maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/ChangeLogScmRequest.java]: 4 char indent

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/changelog/GitChangeLogCommand.java [hunk 7]
```diff
@@ -57,7 +57,20 @@ protected ChangeLogScmResult executeChangeLogCommand( ScmProviderRepository repo
                                                           String datePattern )
         throws ScmException
     {
-        return executeChangeLogCommand( repo, fileSet, null, null, null, datePattern, startVersion, endVersion );
+        return executeChangeLogCommand(
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/changelog/GitChangeLogCommand.java]: Drop this change.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/changelog/GitChangeLogCommand.java [hunk 8]
```diff
@@ -57,7 +57,20 @@ protected ChangeLogScmResult executeChangeLogCommand( ScmProviderRepository repo
...
+        );
+    }
+
+    /** {@inheritDoc} */
+    protected ChangeLogScmResult executeChangeLogCommand( ScmProviderRepository repo, ScmFileSet fileSet,
+                                                          ScmVersion endVersion, String datePattern,
+                                                          boolean startFromRoot )
+            throws ScmException
+    {
+        return executeChangeLogCommand(
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/changelog/GitChangeLogCommand.java]: Please reformat

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/changelog/GitChangeLogCommand.java [hunk 9]
```diff
@@ -89,17 +102,28 @@ protected ChangeLogScmResult executeChangeLogCommand( ChangeLogScmRequest reques
...
             request.getStartDate(), request.getEndDate(), request.getScmBranch(), datePattern, startVersion,
-                endVersion, request.getLimit() );
+                endVersion, request.getLimit(), request.getStartFromRoot() );
     }
 
     protected ChangeLogScmResult executeChangeLogCommand( ScmProviderRepository repo, ScmFileSet fileSet,
                                                           Date startDate, Date endDate, ScmBranch branch,
                                                           String datePattern, ScmVersion startVersion,
                                                           ScmVersion endVersion, Integer limit )
+            throws ScmException
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/changelog/GitChangeLogCommand.java]: 4 char

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/changelog/GitChangeLogCommand.java [hunk 10]
```diff
@@ -89,17 +102,28 @@ protected ChangeLogScmResult executeChangeLogCommand( ChangeLogScmRequest reques
...
+    {
+        return executeChangeLogCommand(
+                repo, fileSet, startDate, endDate, branch, datePattern, startVersion, endVersion, limit, false
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/changelog/GitChangeLogCommand.java]: Please reformat

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/changelog/GitChangeLogCommand.java [hunk 11]
```diff
@@ -136,9 +160,18 @@ public static Commandline createCommandLine( GitScmProviderRepository repository
                                   null );
     }
 
+    static Commandline createCommandLine( GitScmProviderRepository repository, File workingDirectory,
+                                          ScmBranch branch, Date startDate, Date endDate,
+                                          ScmVersion startVersion, ScmVersion endVersion, Integer limit )
+    {
+        return createCommandLine(
+                repository, workingDirectory, branch, startDate, endDate, startVersion, endVersion, limit, false
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/changelog/GitChangeLogCommand.java]: Please reformat

File: maven-scm-api/src/main/java/org/apache/maven/scm/CommandParameter.java (hunk 2 above)
Comment (McFoggy) [reply maven-scm-api/src/main/java/org/apache/maven/scm/CommandParameter.java]: What about a rename of this command parameter to : `FROM_REPOSITORY_START`?

File: maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/ChangeLogScmRequest.java (hunk 5 above)
Comment (McFoggy) [reply maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/ChangeLogScmRequest.java]: sorry I do not understand this one.

File: maven-scm-api/src/main/java/org/apache/maven/scm/CommandParameter.java (hunk 2 above)
Comment (McFoggy) [reply maven-scm-api/src/main/java/org/apache/maven/scm/CommandParameter.java]: also I would then change the method parameter `startFromRoot ` to `fromRepositoryStart`

File: maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/AbstractChangeLogCommand.java (hunk 3 above)
Comment (McFoggy) [reply maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/AbstractChangeLogCommand.java]: was to have a default non throwing exception behavior ; i'll use `throw new ScmException( "Unsupported method for this provider." );` instead

**Comment by McFoggy** [comment]
@michael-o I have taken your remarks into account and rebased to resolve the conflicts.

If you want I can squash the commits.

File: maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/ChangeLogScmRequest.java (hunk 5 above)
Comment (michael-o) [reply maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/ChangeLogScmRequest.java]: My bad, it is already indented with four chars. All is fine.

File: maven-scm-api/src/main/java/org/apache/maven/scm/CommandParameter.java (hunk 2 above)
Comment (michael-o) [reply maven-scm-api/src/main/java/org/apache/maven/scm/CommandParameter.java]: But isn't that the end, not the start? You are giving one ref from which log goes back in time. Doesn't it?

File: maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/AbstractChangeLogCommand.java (hunk 3 above)
Comment (michael-o) [reply maven-scm-api/src/main/java/org/apache/maven/scm/command/changelog/AbstractChangeLogCommand.java]: Just throw `UnsupportedOperationException`.

**Comment by michael-o** [comment]
Please also note my comment in the JIRA ticket.

**Comment by jira-importer** [comment]
Resolve #1116

**Pull request closed without merging**